crates/autopilot/src/run_loop.rs

```diff
@@ -429,6 +467,54 @@ impl RunLoop {
         Ok(())
     }
 
+    /// Tries to find a `settle` contract call with calldata ending in `tag`.
+    ///
+    /// Returns None if no transaction was found within the deadline or the task
+    /// is cancelled.
+    async fn wait_for_settlement_transaction(
+        eth: Arc<infra::Ethereum>,
+        persistence: Arc<infra::Persistence>,
+        tag: &[u8],
+        mut cancellation_rx: watch::Receiver<bool>,
+    ) -> Result<Option<Transaction>> {
+        let start_offset = MAX_REORG_BLOCK_COUNT;
+        let max_blocks_wait = 20;
+        let current = eth.current_block().borrow().number;
+        let start = current.saturating_sub(start_offset);
+        let deadline = current.saturating_add(max_blocks_wait);
+        tracing::debug!(%current, %start, %deadline, ?tag, "waiting for tag");
+        let mut seen_transactions: HashSet<H256> = Default::default();
```

Comment:
This logic is inspired/copied from this PR: https://github.com/cowprotocol/services/pull/1135
Which was already explained in the comment: https://github.com/cowprotocol/services/pull/1135/files#r1088803161